File: java-basic/src/main/java/com/test/java/javatest/동기화/event/service/EventService.java
package com.test.java.javatest.동기화.event.service;

import com.test.java.javatest.동기화.event.domain.Event;
import com.test.java.javatest.동기화.event.repository.EventRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityNotFoundException;

@RequiredArgsConstructor
@Service
public class EventService {

    private final EventRepository eventRepository;

    @Transactional
    public String apply(String eventCode, String userId) {
        // 정원초과여부
        Event event = eventRepository.findByCode(eventCode)
                .orElseThrow(() -> new EntityNotFoundException("이벤트를 찾을 수 없습니다."));

        event.apply();

        return userId + "이벤트 신청 완료";
    }
}

File: kotlin/src/main/kotlin/study/kotlin/event_sample/EventService.kt
package study.kotlin.event_sample

import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class EventService(
    private val applicationEventPublisher: ApplicationEventPublisher,
) {

    @Transactional
    fun success(event: OrderCommitEvent) {
        println("===> success 실행")
        println("...실행중 ....")

        applicationEventPublisher.publishEvent(event)

        println("===> success 종료")
    }

    @Transactional
    fun exception(event: OrderCommitEvent) {
        println("===> exception 실행")
        println("...실행중 ....")

        applicationEventPublisher.publishEvent(event)

        throw IllegalArgumentException("예외 발생!!!")
    }

    @Transactional(noRollbackFor = [IllegalArgumentException::class])
    fun noRollbackTransaction(event: OrderCommitEvent) {
        println("===> noRollbackTransaction 실행")
        println("...실행중 ....")

        applicationEventPublisher.publishEvent(event)

        throw IllegalArgumentException("예외 발생!!!")
    }
}
